projects/rocthrust/test/test_unique_ptr_general.cpp

```diff
@@ -0,0 +1,447 @@
+    }
+    __host__ __device__ A(thrust::device_ptr<bool> dev_p, int num): finished(dev_p), arr(nullptr), n(num) {}
+
+    __host__ __device__ int number() { return n; }
+
+    __host__ __device__ ~A() {
+    #if defined(__HIP_DEVICE_COMPILE__)
+        if(finished)
+            *finished = true; 
+    #endif
+    }
+ 
+    thrust::device_ptr<bool> finished;
+    thrust::device_ptr<int> arr; 
+    int n;
+}; 
+
+template <class T>
+struct A_deleter {
+
+    __host__ void operator()(thrust::device_ptr<T> p) const {
+        A host_copy = *p; 
+        if (host_copy.arr != nullptr) {
+            thrust::device_delete(host_copy.arr, 10);
+        }
+
+        thrust::for_each_n(p, 1, [] __device__(T& x) { x.~T(); });
+        thrust::device_free(p);
+    }
+};
+
+struct GetDeleterTestDeleter
+{
+    void operator()(thrust::device_ptr<int>) const {}
+    void operator()(thrust::device_ptr<int[]>) const {}
+
+    int test() { return 5; }
+    int test() const { return 6; }
+};
+
+TESTS_DEFINE(UniquePtrGeneralTests, NumericalTestsParams);
+
+// Based on libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.special/swap.pass.cpp
+TEST(UniquePtrGeneralTests, TestUniquePtrSwap)
+{
+    SCOPED_TRACE(testing::Message() << "with device_id= " << test::set_device_from_ctest());
+
+    // Test swap for single objects
+    {
+        thrust::unique_ptr<int> p1 = thrust::make_unique<int>(11);
+        thrust::unique_ptr<int> p2 = thrust::make_unique<int>(22);
+
+        int* raw_p1 = p1.get_raw();
+        int* raw_p2 = p2.get_raw();
+
+        p1.swap(p2);
+        ASSERT_EQ(p1.get_raw(), raw_p2);
+        ASSERT_EQ(p2.get_raw(), raw_p1);
+    }
+}
+
+// Thrust-specific test for assignment operator.
+TEST(UniquePtrGeneralTests, TestUniquePtrAsgnMove)
+{
+    SCOPED_TRACE(testing::Message() << "with device_id= " << test::set_device_from_ctest());
+
+    // Move assignment for single objects
+    {
+        thrust::device_ptr<bool> finished = thrust::device_new<bool>(1);
+        *finished = false;
+
+        thrust::unique_ptr<A> p1 = thrust::make_unique<A>(finished, 1);
+        A* raw_p1 = p1.get_raw();
+        {
+            thrust::unique_ptr<A> p2 = thrust::make_unique<A>(nullptr, 2);
+            ASSERT_NE(p1.get_raw(), nullptr);
+            ASSERT_NE(p2.get_raw(), nullptr);
+
+            p2 = std::move(p1);
+
+            ASSERT_EQ(p2.get_raw(), raw_p1);
+            ASSERT_EQ(p1.get_raw(), nullptr);
+        }
+        ASSERT_EQ(*finished, true);
+        thrust::device_delete(finished);
+    }  
+}
+
+// Based on libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.asgn/move.pass.cpp
+TEST(UniquePtrGeneralTests, TestUniquePtrAsgnSelfMove)
+{
+    SCOPED_TRACE(testing::Message() << "with device_id= " << test::set_device_from_ctest());
+
+    // Self-move for single object
+    {
+        thrust::unique_ptr<int> p = thrust::make_unique<int>(1);
+        int* raw_p = p.get_raw();
+        p = std::move(p);
```

Comment:
Clang complains here but I think it's fine to silence the warning here.
```suggestion
        THRUST_DIAG_PUSH
        THRUST_DIAG_SUPPRESS_CLANG("-Wself-move")
        p = std::move(p);
        THRUST_DIAG_POP
```

Reply:
Yes it is, I'll add this